src/Distributed/distributed_incompressible_model.jl

```diff
@@ -0,0 +1,82 @@
+using Oceananigans.Grids
+using Oceananigans.Grids: halo_size
+
+#####
+##### Distributed incompressible model constructor
+#####
+
+function DistributedIncompressibleModel(; architecture, grid, boundary_conditions=nothing, model_kwargs...)
+    my_rank = architecture.my_rank
+    i, j, k = architecture.my_index
+    Rx, Ry, Rz = architecture.ranks
+    my_connectivity = architecture.connectivity
+
+    Nx, Ny, Nz = size(grid)
+    Lx, Ly, Lz = length(grid)
+
+    # Pull out endpoints for full grid.
+    xL, xR = grid.xF[1], grid.xF[Nx+1]
+    yL, yR = grid.yF[1], grid.yF[Ny+1]
+    zL, zR = grid.zF[1], grid.zF[Nz+1]
+
+    # Make sure we can put an integer number of grid points in each rank.
+    # Will generalize in the future.
```

Comment:
You could think about generalizing now? That might help avoid code creeping in that assumes all sub-grids are the same size - maybe? The extra code for this is pretty short (e.g. https://github.com/christophernhill/iap-2021-12.091/blob/e79dfe9dca5441e561cefd65b4c052b1a1dea5a3/step3.py#L36 ).

Reply:
Yes I mentioned this in https://github.com/CliMA/Oceananigans.jl/pull/590#issuecomment-775362818 but those issue comments get buried quite easily. Should have probably opened an issue about it.

Would it be possible to merge PR #590 then make it a priority to support non-equal domain decompositions in a subsequent PR, perhaps even this PR with shallow water model?

I was hoping to merge #590 before it gets too stale but also the incompressible model can only be decomposed in the y-direction limiting our ability to fully test non-equal domain decompositions.

With shallow water model in this PR we could test different x, y, and xy domain decompositions (all non-equal).

Maybe we can come up with some kind of shallow water convergence test then running it with different domain decompositions should produce the same order of convergence?